Source/Classes/AnimatedImages/PINCachedAnimatedImage.m

```diff
@@ -0,0 +1,426 @@
+        [self cleanupFrames];
+    }];
+}
+
+- (void)getKeepRanges:(nonnull out NSRange *)endKeepRangeIn beginningKeepRange:(nonnull out NSRange *)beginningKeepRangeIn
+{
+    __block NSRange endKeepRange;
+    __block NSRange beginningKeepRange;
+    
+    NSUInteger framesToCache = [self framesToCache];
+    
+    [self->_lock lockWithBlock:^{
+        // find the range of frames we want to keep
+        endKeepRange = NSMakeRange(_playhead, framesToCache);
+        beginningKeepRange = NSMakeRange(NSNotFound, 0);
+        if (NSMaxRange(endKeepRange) > _animatedImage.frameCount) {
+            beginningKeepRange = NSMakeRange(0, NSMaxRange(endKeepRange) - _animatedImage.frameCount);
+            endKeepRange.length = _animatedImage.frameCount - _playhead;
+        }
+    }];
+    
+    if (endKeepRangeIn) {
+        *endKeepRangeIn = endKeepRange;
+    }
+    if (beginningKeepRangeIn) {
+        *beginningKeepRangeIn = beginningKeepRange;
+    }
+}
+
+- (void)cleanupFrames
+{
+    NSRange endKeepRange;
+    NSRange beginningKeepRange;
+    [self getKeepRanges:&endKeepRange beginningKeepRange:&beginningKeepRange];
+    
+    [_lock lockWithBlock:^{
+        NSMutableIndexSet *removedFrames = [[NSMutableIndexSet alloc] init];
+        PINLog(@"Checking if frames need removing: %lu", _cachedOrCachingFrames.count);
+        [_cachedOrCachingFrames enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL * _Nonnull stop) {
+            BOOL shouldKeepFrame = NSLocationInRange(idx, endKeepRange);
+            if (beginningKeepRange.location != NSNotFound) {
+                shouldKeepFrame |= NSLocationInRange(idx, beginningKeepRange);
+            }
+            if (shouldKeepFrame == NO) {
+                [removedFrames addIndex:idx];
+                [self->_frameCache removeObjectForKey:@(idx)];
+                PINLog(@"Removing: %lu", (unsigned long)idx);
+            }
+        }];
+        [_cachedOrCachingFrames removeIndexes:removedFrames];
+    }];
+}
+
+- (void)l_cacheFrame:(NSUInteger)frameIndex
+{
+    if ([_cachedOrCachingFrames containsIndex:frameIndex] == NO) {
+        PINLog(@"Requesting: %lu", (unsigned long)frameIndex);
+        [_cachedOrCachingFrames addIndex:frameIndex];
+        _playbackReady++;
+        
+        dispatch_async(_cachingQueue, ^{
+            CGImageRef imageRef = [self->_animatedImage imageAtIndex:frameIndex cacheProvider:self];
+            PINLog(@"Generating: %lu", (unsigned long)frameIndex);
+
+            if (imageRef) {
+                [self->_lock lockWithBlock:^{
+                    [self->_frameCache setObject:(__bridge id _Nonnull)(imageRef) forKey:@(frameIndex)];
+                    self->_playbackReady--;
+                    NSAssert(self->_playbackReady >= 0, @"playback ready is less than zero, something is wrong :(");
+                    
+                    PINLog(@"Frames left: %ld", (long)_playbackReady);
+                    
+                    dispatch_block_t notify = nil;
```

Comment:
Minor, but this could be moved up a couple levels in scope. It's only referenced in one of the inner `ifs` here.